design/specs/foreign-key-constraints.md

```diff
@@ -2,75 +2,169 @@
+- The user hovers over the foreign key reference link below the column name.
+- Information about the constraints applied to the column is displayed in a tooltip.
+- The user clicks on the foreign key reference link below the column name.
+- Constraint settings open for the column.
+
+### Scenario 2b: From table constraint settings
+
+#### Steps for 2b
+
+- A user wants to edit an existing foreign key constraint to change the selected columns or select another referenced table.
+- The user can open a specific foreign key constraint form by clicking on the foreign key link in the column header.
+- If the modified FK constraint causes existing column values to become invalid, the user will receive a warning and be aware of the potential implications of this change, like loss of data or broken views.
+
+## Scenario 3: User Identifies a Column With a Foreign Key Constraint Applied
+
+### Scenario 3a: The foreign key constraint is set to a single column
+
+#### Steps for 3a
+
+- The user opens a table that contains columns with a foreign key constraint applied.
+- The user identifies the columns in two ways:
+  - The user looks at the column header and sees the foreign key indicator. This indicator shows a key icon and the referenced table and column next to it.
+    - The user can click the indicator to display the foreign key settings for the specific column.
+    - The indicator can be hovered over to display a tooltip with information about the foreign key constraints.
+  - The user looks at the column fields and sees that values are added with different styling, a tag-like element with a colored background.
+
+### Scenario 3b: The foreign key constraint is set to multiple columns
+
+#### Steps for 3b
+
+- The user opens a table that contains columns with a multi-column foreign key constraint applied.
+  - The indicator, in this case, looks similar to the single-column one. However, the user will understand that the foreign key constraints are referencing the same table.
+    - The user can tell that a referenced table is the same because they share the same color. The system could use other UI elements instead of color. Still, the idea is to create a visual distinction that can be easily scanned.
+  - The indicator can be hovered over to see additional information. From here, the user will understand that the constraint references more than one column.
+
+## Scenario 4: User Sees a Preview of the Linked Record in a Column With a Foreign Key Constraint Applied
+
+In most cases, the values displayed within a foreign key column won't serve to identify the associated record. A preview of the linked record in a column with a foreign key constraint applied can help the user identify it.
+
+### Scenario 4a: The option for record preview is enabled
+
+#### Steps for 4a
+
+- The user opens a table that contains columns with a foreign key constraint applied.
+  - The column contains the referenced values for each cell. The first five fields are displayed.
```

Comment:
I've been thinking about this more and I think we should have a better algorithm for this than the first five fields. I do think we should show five fields, but I don't think we should rely on the order.
- I think we should prioritize columns that have unique constraints applied.
- We should also prioritize text columns.

We should also allow the user to change which columns we use. They should be able to set (up to) 5 columns to use for search and display of record preview. This should be set for the whole column. We would need a UI for this.

Reply:
Agreed. Especially important if the user can't reorder the columns in a table.

Reply:
+1 on this

Reply:
Agree with folks on this.  Tricky things to consider:
- Some primary key columns would be useful, some very not useful (e.g., the one we create).  Text helps with this, but there may also be other considerations.
- Multi-column unique constraints are potentially useful, but only if all columns are included.
- non-unique columns are _never_ useful unless a unique column is included (that gives info for the user to choose between rows with duplicate entries in non-unique columns).

Reply:
I will add a note to the spec regarding a future issue to define how we help users select appropriate columns. However, at this point, I don't want to block them or prevent them from choosing columns as they might not have set a unique constraint for a column, that in fact, has unique values. This might be a common thing when they start from an imported CSV. It will be related to how we might detect that and recommend that they apply a constraint etc. For now, we could assume that a column with no unique constraint could still be valuable as a search column.